fsspec/utils.py

```diff
@@ -150,13 +173,13 @@ def build_name_function(max_int):
 
     pad_length = int(math.ceil(math.log10(max_int)))
 
-    def name_function(i):
+    def name_function(i: int) -> str:
         return str(i).zfill(pad_length)
 
     return name_function
 
 
-def seek_delimiter(file, delimiter, blocksize):
+def seek_delimiter(file: IO[bytes], delimiter: bytes, blocksize: int) -> bool:
```

Comment:
`file` is "file-like" (i.e., has seek/read), doesn't need to actually be a subclass of anything in particular. Is that what this says?

Reply:
Basically! What we want is a Protocol (https://mypy.readthedocs.io/en/stable/protocols.html), which is a way to formalize duck-typing without requiring subclassing. `IO[bytes]` is a built-in protocol for "file-like objects that return bytes" [0], so it works for anything fulfilling the interface.

[0] In practice though "file-like" is actually ill-defined in practice (e.g. do you want a `ReadOnlyFileLike` to statically disallow `write`?), so we might want to eventually make custom protocols for exactly the subset we need.